docs/architecture/adr-075-rpc-subscription.md

```diff
@@ -0,0 +1,668 @@
+event stream generated by a consensus node. This allows clients to observe the
+state of the consensus network, including details of the consensus algorithm
+state machine, proposals, transaction delivery, and block completion. The
+application may also attach custom key-value attributes to events to expose
+application-specific details to clients.
+
+The event subscription API in the RPC service currently comprises three methods:
+
+1. `subscribe`: A request to subscribe to the events matching a specific
+   [query expression][query-grammar]. Events can be filtered by their key-value
+   attributes, including custom attributes provided by the application.
+
+2. `unsubscribe`: A request to cancel an existing subscription based on its
+   query expression.
+
+3. `unsubscribe_all`: A request to cancel all existing subscriptions belonging
+   to the client.
+
+There are some important technical and UX issues with the current RPC event
+subscription API. The rest of this ADR outlines these problems in detail, and
+proposes a new API scheme intended to address them.
+
+### Issue 1: Persistent connections
+
+To subscribe to a node's event stream, a client needs a persistent connection
+to the node.  Unlike the other methods of the service, for which each call is
+serviced by a short-lived HTTP round trip, subscription delivers a continuous
+stream of events to the client by hijacking the HTTP channel for a websocket.
+The stream (and hence the HTTP request) persists until either the subscription
+is explicitly cancelled, or the connection is closed.
+
+There are several problems with this API:
+
+1. **Expensive per-connection state**: The server must maintain a substantial
+   amount of state per subscriber client:
+
+   - The current implementation uses a [WebSocket][ws] for each active
+     subscriber. The connection must be maintained even if there are no
+     matching events for a given client.
+
+     The server can drop idle connections to save resources, but doing so
+     terminates all subscriptions on those connections and forces those clients
+     to re-connect, adding additional resource churn for the server.
+
+   - In addition, the server maintains a separate buffer of undelivered events
+     for each client.  This is to reduce the dual risks that a client will miss
+     events, and that a slow client could "push back" on the publisher,
+     impeding the progress of consensus.
+
+     Because event traffic is quite bursty, queues can potentially take up a
+     lot of memory. Moreover, each subscriber may have a different filter
+     query, so the server winds up having to duplicate the same events among
+     multiple subscriber queues. Not only does this add memory pressure, but it
+     does so most at the worst possible time, i.e., when the server is already
+     under load from high event traffic.
+
+2. **Operational access control is difficult**: The server's websocket
```

Comment:
I have one more argument here.

(Disclaimer: I'm not a fan of CORS in our context as it irrelevant for the Tendermint API and I think CORS should be configured to be all open to get the same behaviour for browser clients as command line clients.)

But other people might see value in it as it can serve as a request filter. Using ws instead of HTTP on a given RPC endpoint is a way to bypass CORS restrictions.

Reply:
That's an interesting point. 

I believe the default CORS policy is "allow-all", unless the operator configures it otherwise. But that would also be _very_ easy to work around in a reverse proxy, if it causes any real problems.

Do you know of any specific configurations where CORS policy is a blocker (or more than just a passing hassle)?

Reply:
CORS is a blocker all the time because ops people have to open everything up explicitly. Then the endpoints are tested from command line and [poor newcomer dev open support request at CosmJS](https://github.com/cosmos/cosmjs/issues/993).

The reasons are probably:
- Opening up CORS to allow all requests is not trivial
- It sounds all scary if you don't know what CORS was designed for
- Cosmos SDK calls it unsafe
  ```
  # EnableUnsafeCORS defines if CORS should be enabled (unsafe - use it at your own risk).
  enabled-unsafe-cors = true
  ```
  
I think most RPC endpoint do not have the CORS config they should have. Using wss:// is usually a nice way to get that out of the way.

Reply:
Interesting. It doesn't really make sense for the Tendermint HTTP server to restrict CORS policy by default; if an operator wants anything other than "allow-all" they should have to say so.

That said, given the other problems with the websocket, I think we should look for better ways to address CORS issues, rather than keeping it as a workaround. 🙂

Reply:
Right. To frame it positive you can say that access restriction for browsers through the use of CORS becomes easier to maintain when using HTTP only.

Reply:
I think what would even be more valuable is educational content about what the team's CORS recommendation is and why it is safe in this context.

I did not check the latest implementation and the defaults. You need some more headers. Also I'm not sure if Cosmos SDK overrides this when Tendermint is not running in its own process.

Reply:
> I think what would even be more valuable is educational content about what the team's CORS recommendation is and why it is safe to open up for this API.

Yes, I agree that more documentation would probably help in any case. 

But from an implementation perspective, the server that provides the UI is what needs to specify CORS rules. Apart from the list-of-endpoints default, the node's RPC server doesn't serve any HTML or other browser-specific resources.

It probably makes sense to have the RPC server _report_ an allow-origin policy, and maybe respond affirmatively to preflights—just so the browser doesn't complain. I think, though, that any other policy settings should probably _not_ be handled by the node itself.

I'm not sure about how the SDK handles this, though.

> I did not check the latest implementation and the defaults. You need some more headers.

I think you only need more headers if you want to set a non-trivial policy (e.g., to specify which other headers are allowed in the XO request). But as you say, much depends on what you're trying to do.